File: website/helpers/parameter_overview.R
library(dplyr)
library(glue)
library(jsonlite)
library(readr)
library(stringr)

`%||%` <- function(x, y) {
  if (is.null(x)) y else x
}

find_parameter_manifest_root <- function(start_dir = ".") {
  current_dir <- normalizePath(start_dir, winslash = "/", mustWork = TRUE)

  repeat {
    if (file.exists(file.path(current_dir, "cfg_pipeline_parameters.tsv"))) {
      return(current_dir)
    }

    parent_dir <- dirname(current_dir)
    if (identical(parent_dir, current_dir)) {
      stop("Could not find cfg_pipeline_parameters.tsv in current or parent directories.", call. = FALSE)
    }
    current_dir <- parent_dir
  }
}

read_parameter_manifest <- function(manifest_file) {
  manifest <- readr::read_tsv(
    manifest_file,
    col_types = readr::cols(
      .default = readr::col_character(),
      allow_missing_after_inheritance = readr::col_logical()
    ),
    na = character(),
    show_col_types = FALSE
  )

  names(manifest) <- stringr::str_trim(names(manifest))
  dplyr::mutate(manifest, dplyr::across(where(is.character), stringr::str_trim))
}

parameter_overview_data <- function(manifest, scope = NULL) {
  parameter_tibble <- manifest |>
    dplyr::filter(is.null(.env$scope) | .data$scope %in% .env$scope) |>
    dplyr::mutate(
      status = dplyr::case_when(
        .data$default_value != "NULL" ~ "Defaulted",
        !.data$allow_missing_after_inheritance ~ "Required",
        TRUE ~ "Optional"
      )
    ) |>
    dplyr::select(
      scope, param_name, short_name, data_type, cardinality, default_value,
      allowed_values, examples, part_of, description, status
    )

  if (nrow(parameter_tibble) == 0) {
    stop("No parameters found for manifest scope: ", scope, call. = FALSE)
  }
  list(parameters = parameter_tibble)
}

render_parameter_overview_fragment <- function(
  overview_data,
  scope,
  search_placeholder = NULL
) {
  search_placeholder <- search_placeholder %||% "Search names, descriptions or values"
  data_json <- jsonlite::toJSON(overview_data, auto_unbox = TRUE, null = "null")
  # Prevent a manifest value from closing the inert JSON script element.
  data_json <- gsub("<", "\\u003c", data_json, fixed = TRUE)
  script <- paste(readLines(file.path(find_parameter_manifest_root(), "website", "helpers", "parameter_overview.js")), collapse = "\n")
  search_placeholder <- htmltools::htmlEscape(search_placeholder, attribute = TRUE)

  glue::glue(
'<div class="parameter-overview" data-parameter-overview>
<div class="parameter-workflows" role="tablist" aria-label="Workflow"></div>
<div class="parameter-overview-toolbar">
<label class="parameter-search-control"><span>Search this workflow</span>
<input class="parameter-overview-search" type="search" placeholder="{search_placeholder}"></label>
<button type="button" class="parameter-reset">Clear search</button>
</div>
<p class="parameter-overview-help">Required: supply a value directly or through inheritance. Defaulted: a non-null default is provided. Optional: may remain unset. Within each group, parameters are sorted by short name.</p>
<p class="parameter-overview-summary" aria-live="polite" aria-atomic="true"></p>
<div class="parameter-panel" id="parameter-panel" role="tabpanel">
<div class="parameter-section-list"></div>
<p class="parameter-empty-state" hidden>No parameters match in this workflow. Clear the search or choose another tab.</p>
</div>
<noscript><p>Enable JavaScript to browse this reference, or read <a href="https://github.com/koefoeden/multiomeR/blob/main/cfg_pipeline_parameters.tsv">the parameter manifest</a>.</p></noscript>
<script type="application/json" class="parameter-overview-data">{data_json}</script>
</div>
<script>{script}</script>'
  )
}

emit_parameter_overview <- function(
  scope,
  manifest_file = file.path(find_parameter_manifest_root(), "website", "data", "public_defaults", "cfg_pipeline_parameters.tsv"),
  search_placeholder = NULL
) {
  manifest <- read_parameter_manifest(manifest_file)
  overview_data <- parameter_overview_data(manifest, scope)
  cat("```{=html}\n", render_parameter_overview_fragment(overview_data, scope, search_placeholder), "\n```\n", sep = "")
}

render_parameter_overview_document <- function(overview_data, scope, search_placeholder = NULL) {
  scope_label <- if (is.null(scope)) "All" else stringr::str_to_sentence(stringr::str_replace_all(scope, "_", " "))
  styles <- paste(readLines(file.path(find_parameter_manifest_root(), "website", "styles.css")), collapse = "\n")
  fragment <- render_parameter_overview_fragment(overview_data, scope, search_placeholder)

  glue::glue(
'<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>{scope_label} parameters | multiomeR</title>
  <style>{styles}</style>
</head>
<body class="parameter-overview-standalone">
<main>
<h1>{scope_label} parameters</h1>
<p>Search settings across the pipeline and optional modules. Open a parameter for details or use its permalink to share it.</p>
{fragment}
</main>
</body>
</html>'
  )
}

render_parameter_overview_file <- function(
  scope = NULL,
  manifest_file = file.path(find_parameter_manifest_root(), "website", "data", "public_defaults", "cfg_pipeline_parameters.tsv"),
  output_file = NULL,
  search_placeholder = NULL
) {
  repo_root <- find_parameter_manifest_root()
  if (is.null(output_file)) {
    output_file <- file.path(repo_root, "website", "parameters.html")
  }

  manifest <- read_parameter_manifest(manifest_file)
  overview_data <- parameter_overview_data(manifest, scope)
  html <- render_parameter_overview_document(overview_data, scope, search_placeholder)

  dir.create(dirname(output_file), recursive = TRUE, showWarnings = FALSE)
  writeLines(html, output_file, useBytes = TRUE)
  message("Wrote ", normalizePath(output_file, winslash = "/", mustWork = FALSE))
  invisible(output_file)
}

File: website/helpers/parameter_overview.js
(() => {
  const widget = document.currentScript.previousElementSibling;
  const parameters = JSON.parse(widget.querySelector('.parameter-overview-data').textContent).parameters;
  const search = widget.querySelector('input[type="search"]');
  const tabs = widget.querySelector('[role="tablist"]');
  const panel = widget.querySelector('[role="tabpanel"]');
  const list = widget.querySelector('.parameter-section-list');
  const summary = widget.querySelector('.parameter-overview-summary');
  const empty = widget.querySelector('.parameter-empty-state');
  const workflows = [
    ['aggregation', 'Main workflow'],
    ['differential_analyses', 'Differential analyses'],
    ['genetic_enrichment', 'Genetic enrichment'],
    ['peak_gene_correlation', 'Peak–gene correlation']
  ].filter(([scope]) => parameters.some(parameter => parameter.scope === scope));
  let workflow = workflows[0][0];
  const expanded = new Set();
  const escape = value => String(value ?? '').replaceAll('&', '&amp;').replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;').replaceAll('"', '&quot;').replaceAll("'", '&#39;');
  const searchText = parameter => Object.values(parameter).join(' ').toLowerCase();
  tabs.innerHTML = workflows.map(([scope, label]) => `<button type="button" role="tab" class="parameter-workflow" id="workflow-${scope}" data-scope="${scope}" aria-controls="parameter-panel">${label}<span>${parameters.filter(parameter => parameter.scope === scope).length}</span></button>`).join('');
  const buttons = [...tabs.querySelectorAll('button')];

  function renderRow(parameter) {
    const defaultValue = parameter.status === 'Required' ? 'Supply a value' : parameter.default_value;
    const detail = (label, value, code = true) => value ? `<div class="parameter-detail"><dt>${escape(label)}</dt><dd>${code ? `<code>${escape(value)}</code>` : escape(value)}</dd></div>` : '';
    return `<details class="parameter-row" id="${escape(parameter.param_name)}" tabindex="-1" data-param="${escape(parameter.param_name)}"${expanded.has(parameter.param_name) ? ' open' : ''}>
      <summary class="parameter-head">
        <span class="parameter-name-wrap">
          <span class="parameter-title"><strong>${escape(parameter.short_name)}</strong><span class="parameter-status ${parameter.status.toLowerCase()}">${parameter.status}</span></span>
          <code class="parameter-name">${escape(parameter.param_name)}</code>
        </span>
        <span class="parameter-default"><span>Default</span><code>${escape(defaultValue)}</code></span>
        <span class="parameter-description">${escape(parameter.description)}</span>
      </summary>
      <p><a href="#${encodeURIComponent(parameter.param_name)}" class="parameter-permalink">Link to this parameter</a></p>
      <dl class="parameter-details">
        ${detail('Type', `${parameter.data_type} · ${parameter.cardinality}`, false)}
        ${detail('Allowed values', parameter.allowed_values || 'Any value matching the type', Boolean(parameter.allowed_values))}
        ${detail('Example', parameter.examples)}
        ${detail('Used by', parameter.part_of, false)}
      </dl>
    </details>`;
  }

  function render() {
    const words = search.value.trim().toLowerCase().split(/\s+/).filter(Boolean);
    const available = parameters.filter(parameter => parameter.scope === workflow);
    const visible = available.filter(parameter => words.every(word => searchText(parameter).includes(word)))
      .sort((a, b) => a.short_name.localeCompare(b.short_name, 'en', {sensitivity: 'base'}) || a.param_name.localeCompare(b.param_name));
    buttons.forEach(button => {
      const active = button.dataset.scope === workflow;
      button.setAttribute('aria-selected', String(active));
      button.tabIndex = active ? 0 : -1;
    });
    panel.setAttribute('aria-labelledby', `workflow-${workflow}`);
    list.innerHTML = ['Required', 'Defaulted', 'Optional'].map(status => {
      const rows = visible.filter(parameter => parameter.status === status);
      return rows.length ? `<section class="parameter-section" data-status="${status}"><h2>${status}<span>${rows.length}</span></h2>
        <div class="parameter-list">${rows.map(renderRow).join('')}</div></section>` : '';
    }).join('');
    summary.textContent = `${workflows.find(([scope]) => scope === workflow)[1]}: ${visible.length} of ${available.length} parameters`;
    empty.hidden = visible.length !== 0;
    list.querySelectorAll('details').forEach(row => row.addEventListener('toggle', () => {
      if (row.open) expanded.add(row.dataset.param); else expanded.delete(row.dataset.param);
    }));
  }

  function navigate() {
    let fragment;
    try { fragment = decodeURIComponent(window.location.hash.slice(1)); } catch { return; }
    const parameter = parameters.find(parameter => parameter.param_name === fragment);
    const scope = fragment.startsWith('workflow=') ? fragment.slice('workflow='.length) : null;
    if (parameter) workflow = parameter.scope;
    else if (workflows.some(([value]) => value === scope)) workflow = scope;
    else if (!fragment) workflow = workflows[0][0];
    else return;
    search.value = '';
    if (parameter) expanded.add(parameter.param_name);
    render();
    if (parameter) {
      const row = document.getElementById(parameter.param_name);
      row.focus({preventScroll: true});
      row.scrollIntoView({block: 'center'});
    }
  }
  buttons.forEach(button => button.addEventListener('click', () => {
    const hash = `#workflow=${button.dataset.scope}`;
    if (window.location.hash === hash) navigate(); else window.location.hash = hash;
  }));
  tabs.addEventListener('keydown', event => {
    const index = buttons.indexOf(event.target);
    if (index < 0) return;
    const next = {ArrowRight: (index + 1) % buttons.length, ArrowLeft: (index + buttons.length - 1) % buttons.length, Home: 0, End: buttons.length - 1}[event.key];
    if (next === undefined) return;
    event.preventDefault();
    buttons[next].focus();
    buttons[next].click();
  });
  search.addEventListener('input', render);
  widget.querySelector('.parameter-reset').addEventListener('click', () => {
    search.value = ''; render(); search.focus();
  });
  window.addEventListener('hashchange', navigate);
  widget.addEventListener('click', event => {
    if (event.target.closest('.parameter-permalink')?.hash === window.location.hash) navigate();
  });
  render();
  navigate();
})();
